sos/upload/__init__.py

```diff
@@ -0,0 +1,325 @@
+        checks = list(self.upload_targets.values())
+        for upload_target in self.upload_targets.values():
+            checks.remove(upload_target)
+            if upload_target.check_distribution():
+                cname = upload_target.__class__.__name__
+                self.ui_log.debug(f"Installation matches {cname}, checking for"
+                                  " upload targets")
+                self.upload_target = upload_target
+                self.upload_name = upload_target.name()
+                self.ui_log.info(
+                    f'Upload target set to {self.upload_name}')
+                break
+
+    def load_upload_targets(self):
+        """Loads all upload targets supported by the local installation
+        """
+        import sos.upload.targets
+        package = sos.upload.targets
+        supported_upload_targets = {}
+        upload_targets = self._load_modules(package, 'targets')
+        for upload_target in upload_targets:
+            target_class = upload_target[1](
+                            parser=self.parser,
+                            args=self.args,
+                            cmdline=self.cmdline
+                            )
+            target_class.set_commons(self.get_commons())
+            supported_upload_targets[upload_target[0]] = target_class
+        return supported_upload_targets
+
+    @classmethod
+    def _load_modules(cls, package, submod):
+        """Helper to import upload targets"""
+        modules = []
+        for path in package.__path__:
+            if os.path.isdir(path):
+                modules.extend(cls._find_modules_in_path(path, submod))
+        return modules
+
+    @classmethod
+    def _find_modules_in_path(cls, path, modulename):
+        """Given a path and a module name, find everything that can be imported
+        and then import it
+
+            path - the filesystem path of the package
+            modulename - the name of the module in the package
+
+        E.G. a path of 'targets', and a modulename of 'redhat' equates to
+        importing sos.upload.targets.redhat
+        """
+        modules = []
+        if os.path.exists(path):
+            for pyfile in sorted(os.listdir(path)):
+                if not pyfile.endswith('.py'):
+                    continue
+                if '__' in pyfile:
+                    continue
+                fname, _ = os.path.splitext(pyfile)
+                modname = f'sos.upload.{modulename}.{fname}'
+                modules.extend(cls._import_modules(modname))
+        return modules
+
+    @classmethod
+    def _import_modules(cls, modname):
+        """Import and return all found classes in a module"""
+        mod_short_name = modname.split('.')[2]
+        try:
+            module = __import__(modname, globals(), locals(), [mod_short_name])
+        except ImportError as e:
+            raise e
+        modules = inspect.getmembers(module, inspect.isclass)
+        for mod in modules.copy():
```

Comment:
Good usage of `copy` to prevent removing items from a list you are traversing through :)